File: Ruby/Hasta Entrega 3/Irrgarten/Irrgarten/monster.rb
module Irrgarten
  class Monster
    @@initial_health = 5
    @@unset_pos = -1

    def initialize(name,intelligence,strength)
      @name = name
      @intelligence = intelligence
      @strength = strength
      @health = @@initial_health
      @row = @@unset_pos
      @col = @@unset_pos
    end

    # Devuelve true si el monstruo está muerto
    def dead?
      @health <= 0
    end

    # Devuelve la cantidad de daño que hace el mosntruo
    def attack
      Dice.intensity(@strength)
    end

    # Metodo completado
    def defend(received_attack)
      is_dead = dead

      unless is_dead
        defensive_energy = Dice.intensity(@intelligence)

        if defensive_energy < received_attack
          got_wounded
          is_dead = dead
        end
      end

      is_dead
    end

    # Setter
    def set_pos(row,col)
      @row = row
      @col = col
    end

    #Sobrecarga del método to_string
    def to_s
      "M[#{@name}, #{@intelligence}, #{@strength}, #{@health}, #{@row}, #{@col}]"
    end

    private
    #Decrementa la salud en 1
    def got_wounded
      if @health > 0
        @health -= 1
      end
    end

  end
end

File: Java/Hasta Entraga 3/irrgarten/src/irrgarten/Monster.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

/**
 *
 * @author alejandro
 */
public class Monster {
    // Atributos de clases estáticos
    private static final int INITIAL_HEALTH = 5;
    private static final int UNSET_POS = -1;
    
    // Atributos de instancia
    private String name;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;

    // Constructor de la clase Monster
    public Monster(String name, float intelligence, float strength) {
        this.name = name;
        this.intelligence = intelligence;
        this.strength = strength;
        this.health = INITIAL_HEALTH;
        this.row = UNSET_POS;
        this.col = UNSET_POS;
    }
    
    // Devuelve true si el monstruo está muerto
    public boolean dead(){
        return health <= 0;
    }
    
    // Devuelve la cantidad de daño que hace el monstruo
    public float attack(){
        return Dice.intensity(strength);
    }
    
    // Devuelve la cantidad de daño del que se defienfe el monstruo
    public boolean defend(float receivedAttack){
        boolean isDead = dead();
        
        if(!isDead){
            float devensiveEnergy = Dice.intensity(intelligence);
            if(devensiveEnergy < receivedAttack){
                gotWounded();
                isDead = dead();
            }
        }
        
        return isDead;
    }
    
    // Setter
    public void setPos(int row, int col){
        this.row = row;
        this.col = col;
    }
    
    @Override
    public String toString(){
        return "M[" + name + " , " + intelligence + " , " +
                strength + " , " + health + " , " + row +
                " , " + col + " ]";
    }
    
    // Decrementa en 1 la salud
    private void gotWounded(){
        if(this.health > 0)
            this.health--; 
    }
}

